<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />



<meta name="date" content="2020-01-20" />

<title>Reproducible scripting, AlphaSimR, Homeworks</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Lab1</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
<li>
  <a href="Lab1_ProgrammingAlphaSimR.html">Lab1</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/jeanlucj/Lab1_7170">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Reproducible scripting, AlphaSimR, Homeworks</h1>
<h4 class="author">Jean-Luc Jannink and Kelly Robbins</h4>
<address class="author_afil">
USDA-ARS / Cornell<br><h4 class="date">January 20, 2020</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-01-20
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>Lab1_7170/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200120code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200120)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200120code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200120)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomjeanlucjLab17170treec641e33cbccf7460e207a8fb644c655479b4bc87targetblankc641e33a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/jeanlucj/Lab1_7170/tree/c641e33cbccf7460e207a8fb644c655479b4bc87" target="_blank">c641e33</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomjeanlucjLab17170treec641e33cbccf7460e207a8fb644c655479b4bc87targetblankc641e33a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  .addToReadMe.txt

Unstaged changes:
    Deleted:    data/FileStructure.jpg
    Deleted:    data/RcodeInDescription.jpg

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/c641e33cbccf7460e207a8fb644c655479b4bc87/analysis/Lab1_ProgrammingAlphaSimR.Rmd" target="_blank">c641e33</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
<td>
Moving linked figures to see if they will appear in GitHub.io
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jeanlucj/Lab1_7170/d91a4ffd6c41d77658da6a9efc7d8f42c1e97555/docs/Lab1_ProgrammingAlphaSimR.html" target="_blank">d91a4ff</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/9d95adde55e31a5804fe2f893de73e1dae128329/analysis/Lab1_ProgrammingAlphaSimR.Rmd" target="_blank">9d95add</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
<td>
Touch ups on the main document
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/jeanlucj/Lab1_7170/602d7eb7d861d46218767fb5b678276a51545c4c/docs/Lab1_ProgrammingAlphaSimR.html" target="_blank">602d7eb</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/cb851516e6e309e7563b5e44db501f196ad33163/analysis/Lab1_ProgrammingAlphaSimR.Rmd" target="_blank">cb85151</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
<td>
Adding all the untracked files
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>
<div id="preliminaries" class="section level1">
<h1>Preliminaries</h1>
<div id="pre-preparation" class="section level2">
<h2>Pre-preparation</h2>
<p>If you rate your familiarity with R as “low” please go to <a href="http://www.datacarpentry.org/R-genomics/">Data Carpentry: R for Genomics</a> <em>in advance</em> of the first class and do the following:</p>
<ul>
<li>Work through the material between <em>at minimum</em> chapters 1 (“Before We Start”) and 4 (“Data frames”), but preferably also 5 (“The dplyr package”).</li>
<li>This website includes instructions for installing R and Rstudio and <em>basic</em> instruction on how to interact with R.</li>
<li>Please install the <code>AlphaSimR</code>, <code>workflowr</code>, and <code>tidyverse</code> packages.</li>
</ul>
</div>
<div id="learning-objectives" class="section level2">
<h2>Learning objectives</h2>
<ol style="list-style-type: decimal">
<li>Reproducible scripting</li>
</ol>
<ol style="list-style-type: lower-roman">
<li>A reproducible analysis workflow includes a predictable input file and data structure, and outputs that are described, interpreted, and put in the context of the overall project in English.<br />
</li>
<li>The audience of this workflow is akin to someone who might be reviewing a manuscript derived from the work. The <strong>most important audience is yourself, six months later</strong>, or your close collaborators (e.g., lab group) who may need to carry on the work after you move on.<br />
</li>
<li>Whether you like it or not, you are a computational biologist. Lab biology experiments need to be reproducible. Computational biology analyses need to be reproducible.<br />
</li>
</ol>
<ol start="2" style="list-style-type: decimal">
<li>R markdown and workflowr<br />
</li>
</ol>
<ol style="list-style-type: lower-roman">
<li>R markdown language. It allows you to mingle script with formatted text, with script outputs. Note that Python and c++ scripts can be incorporated into R markdown. Help pages on rmarkdown:<br />
<a href="https://rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">Cheat Sheet</a><br />
<a href="https://rmarkdown.rstudio.com/authoring_basics.html">Authoring Basics</a><br />
<a href="https://annakrystalli.me/rrresearch/05_literate-prog.html">In Literate Programming</a><br />
</li>
<li><code>workflowr</code> aims to make it easier to follow a coding workflow that will increase the communicability and reproducibility of your code, and constrain you somewhat to following that workflow. The package sets up a standard directory file structure and populates it with useful initial R markdown (.Rmd) files.<br />
</li>
<li><code>workflowr</code> also sets up the directory as a git version control repository. We will (probably) not teach git, but we encourage you all delving into it. Assume for the purpose of this class that you can use <code>workflowr</code> just to set up the file structure. <img src="./FileStructure.jpg" alt="Standard script directory organization" /><br />
There <a href="https://cran.r-project.org/web/packages/workflowr/vignettes/wflow-01-getting-started.html">is more</a> you can do with the package, and it’s worth checking it out (and its extensive documentation!)<br />
</li>
</ol>
<ol start="3" style="list-style-type: decimal">
<li>AlphaSimR<br />
</li>
</ol>
<ol style="list-style-type: lower-roman">
<li>AlphaSimR is a package to simulate breeding populations and tasks. It is not completely intuitive nor completely well-documented. We will get started with it here. It should help for future homework<br />
</li>
</ol>
<ol start="4" style="list-style-type: decimal">
<li>Homework<br />
</li>
</ol>
<ol style="list-style-type: lower-roman">
<li>Labs, including this one, will be coupled with homework. One possibility for homework will be to do it in a <code>workflowr</code>-created directory, and hand in that zipped directory, or, if you go further and create a webpage out of it, just send in the URL… If all is well, this lab has a webpage here:<br />
<a href="https://jeanlucj.github.io/Lab1_7170/Lab1_ProgrammingAlphaSimR.html">Lab1 GitHub webpage</a><br />
I am noting, however, that on github.io, it is not rendering the within-directory-linked figures.</li>
</ol>
</div>
<div id="document-explanation" class="section level2">
<h2>Document explanation</h2>
<p>There are three components to this document: 1. Discussing “Reproducible Programming” and various practices and packages that can help out.<br />
2. Writing a simple <code>AlphaSimR</code> script and describing it.<br />
3. Setting up a homework.<br />
In principle, we should cover the ideas of the homework in class. You will then make a clean rmarkdown script to show the ideas. The rmarkdown script can be embedded in a <code>workflowr</code> directory if you want. The questions to answer through the lab and homework will be indented text.</p>
<blockquote>
<p>This is indented text</p>
</blockquote>
</div>
<div id="further-background" class="section level2">
<h2>Further background</h2>
<ul>
<li>R markdown is very good for documenting <strong>scripts</strong>, but less so <strong>programs</strong>. If code is linear and fairly simple, I call it a script. If code has loops, is potentially multi-purpose, and defines functions, I call it a program. <code>workflowr</code> provides a <code>code</code> directory for programs, and an <code>analysis</code> directory for scripts. Of course, scripts can (should) refer to functions, etc., in the <code>code</code> directory. Any raw data can go in <code>data</code>. Script outputs can go in <code>output</code>. Final figures and so on can go in <code>docs</code>.<br />
</li>
<li>If you write a series of functions that you will use repeatedly, it’s probably worth <a href="http://r-pkgs.had.co.nz/">making a package</a> out of them. That is not trivial, but it’s less difficult than it sounds. You do not have to submit your package to <strong>CRAN</strong>, but can just use it internally. The documentation of functions that goes along with making a package is very helpful over time.<br />
</li>
<li>If you write a program that you imagine will develop over time, learn version control (probably “git”. I don’t know any other version control…), <a href="https://guides.github.com/activities/hello-world/">here</a> or <a href="https://www.atlassian.com/git">here</a>. Note that a public repository like github can be quite useful for making your data available once you publish your research. Note too that <code>workflowr</code> commands wrap a number of git commands to simplify the work flow, particularly if you are working in the RStudio environment.<br />
</li>
<li>Here are some useful articles:
<ol style="list-style-type: lower-roman">
<li><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Ten Simple Rules for Reproducible Computational Research</a><br />
</li>
<li><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005510">Good enough practices in scientific computing</a></li>
<li><a href="https://f1000research.com/articles/8-1749/v1">Creating and sharing reproducible research code the workflowr way</a></li>
</ol></li>
</ul>
<hr />
</div>
</div>
<div id="simulating-some-classical-results" class="section level1">
<h1>Simulating some classical results</h1>
<p>This script uses <code>AlphaSimR</code> to simulate the “outbreak of variation” that arises when you self-fertilize a hybrid.</p>
<div id="loading-packages" class="section level2">
<h2>Loading packages</h2>
<p>If your script depends on external packages, load them at the beginning. This shows users early on what the script dependencies are.</p>
<pre class="r"><code>ip &lt;- installed.packages()
packages_used &lt;- c(&quot;AlphaSimR&quot;, &quot;tidyverse&quot;, &quot;workflowr&quot;)

for (package in packages_used){
  isInstalled &lt;- package %in% rownames(ip)
  if (!isInstalled){
    stop(paste(&quot;ERROR: you need to install the&quot;, package, &quot;package&quot;))
  } else{
    library(package, character.only=T)
  }
}#END packages_used</code></pre>
<pre><code>Loading required package: R6</code></pre>
<pre><code>── Attaching packages ─────────────────────────────────── tidyverse 1.3.0 ──</code></pre>
<pre><code>✓ ggplot2 3.2.1     ✓ purrr   0.3.3
✓ tibble  2.1.3     ✓ dplyr   0.8.3
✓ tidyr   1.0.0     ✓ stringr 1.4.0
✓ readr   1.3.1     ✓ forcats 0.4.0</code></pre>
<pre><code>── Conflicts ────────────────────────────────────── tidyverse_conflicts() ──
x dplyr::filter() masks stats::filter()
x dplyr::lag()    masks stats::lag()
x dplyr::mutate() masks AlphaSimR::mutate()</code></pre>
<p>A first thing to notice from the output of loading tidyverse is the conflicts report. In particular, for example, two packages (<code>dplyr</code> and <code>stats</code>) have been loaded that have a function called <code>filter</code>. Since dplyr was loaded <em>after</em> stats, if you use the function <code>filter</code>, it will go to the dplyr version by default. It is dangerous to rely on what order packages have been loaded to determine which <code>filter</code> function you get. R syntax to prevent ambiguity is to write either <code>dplyr::filter</code> or <code>stats::filter</code>. Using that syntax will make your code more reproducible.</p>
</div>
<div id="document-packages-used" class="section level2">
<h2>Document packages used</h2>
<p>This chunk creates a “README.md” file that documents the packages and versions used for future reference.</p>
<pre class="r"><code>source(&quot;code/addToREADME.R&quot;)
addToREADME(paste0(&quot;## &quot;, rmarkdown::metadata$title), append=F)
addToREADME(c(date(), &quot;&quot;))

packages_used &lt;- as_tibble(ip[packages_used, c(&quot;Package&quot;, &quot;Version&quot;, &quot;Built&quot;)])
# Clunky way to add the packages at the end
readme_file &lt;- &quot;.addToReadMe.txt&quot;
write_tsv(packages_used, readme_file)</code></pre>
<p>Hyperlink the <a href="../README.md">README file</a> to the report. That makes is easy to find.</p>
</div>
<div id="set-random-seed" class="section level2">
<h2>Set random seed</h2>
<p>AlphaSimR generates many random numbers (e.g., to simulate Mendelian random segregation). If you want the result of the analysis to come out the same each time (there are pros and cons) you need to set the random seed. Note that <code>workflowr</code> does this by default. If you are not using that package, then be explicit.</p>
<pre class="r"><code>random_seed &lt;- 45678
set.seed(random_seed)
addToREADME(paste(&quot;The random seed is&quot;, random_seed))</code></pre>
</div>
<div id="script-parameters" class="section level2">
<h2>Script parameters</h2>
<p>If the behavior of your script depends on parameters that you set, initialize them early on.</p>
<pre class="r"><code>nFounders &lt;- 100
nChr &lt;- 10 # Number of chromosomes
segSites &lt;- 1000 # Number of segregating sites _per chromosome_
nQTL &lt;- 10 # Vary this parameter to get oligo- versus poly- genic traits</code></pre>
</div>
<div id="executable-descriptions" class="section level2">
<h2>Executable descriptions</h2>
<p>It is easy for text descriptions around your code to become out of sync with the code. Because the descriptions are ignored by the computer during execution, they may say one thing while the script says another. That becomes confusing later when you revisit the script. One way to avoid that is “executable descriptions”. For example, rather than setting parameters in a chunk of code, you can set them “inline” in R markdown:<br />
The initial number of founders <code>nFounders</code> = 100<br />
The number of chromosomes for the species is <code>nChr</code> = 10<br />
Each chromosome has <code>segSites</code> = 1000 segregating sites<br />
Each chromosome has <code>nQTL</code> = 10 QTL (i.e. segregating causal variants).<br />
<img src="./RcodeInDescription.jpg" alt="The R markdown for that looks like this" /></p>
</div>
<div id="parameters-to-readme" class="section level2">
<h2>Parameters to README</h2>
<p>This is starting to look <strong>VERY</strong> redundant, but it can’t hurt to have the information in multiple places. Chunks of code like this do not need to be included in the report. To exclude them use the “include=FALSE” option in the chunk.</p>
<pre class="r"><code>addToREADME(c(
  paste(&quot;The number of founders for the simulation is&quot;, nFounders), 
  paste(&quot;The number of chromosomes is&quot;, nChr), 
  paste(&quot;There are&quot;, nQTL, &quot;QTL per chromosome&quot;),
  paste(&quot;There are&quot;, segSites, &quot;segregating sites per chromosome&quot;)
  ))
addToREADME(c(&quot;&quot;, &quot;The packages used in this script are:&quot;))
addToREADME(readLines(readme_file))</code></pre>
</div>
<div id="alphasimr-populations" class="section level2">
<h2>AlphaSimR populations</h2>
<p>The basic object of <code>AlphaSimR</code> is the population. To make founders, you first make founder haplotypes from a coalescent simulation, then you define simulation parameters that will link their genetic variation to phenotypic variation, then you make a first diploid population from the founder haplotypes.</p>
<pre class="r"><code># Create haplotypes for founder population of outbred individuals
# Note: default effective population size for runMacs is 100
founderHaps &lt;- runMacs(nInd=nFounders, nChr=nChr, segSites=segSites)

# New global simulation parameters from founder haplotypes
SP &lt;- SimParam$new(founderHaps)
# Additive and dominance trait architecture
# By default, the genetic variance will be 1
SP$addTraitA(nQtlPerChr=nQTL)

# Create a new population of founders
founders &lt;- newPop(founderHaps, simParam=SP)
str(founders)</code></pre>
<pre><code>Formal class &#39;Pop&#39; [package &quot;AlphaSimR&quot;] with 16 slots
  ..@ id     : chr [1:100] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; ...
  ..@ mother : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ father : chr [1:100] &quot;0&quot; &quot;0&quot; &quot;0&quot; &quot;0&quot; ...
  ..@ gender : chr [1:100] &quot;&quot; &quot;&quot; &quot;&quot; &quot;&quot; ...
  ..@ nTraits: int 1
  ..@ gv     : num [1:100, 1] -0.139 0.554 -0.103 -0.49 1.3 ...
  ..@ pheno  : num [1:100, 1] NA NA NA NA NA NA NA NA NA NA ...
  ..@ ebv    : num[1:100, 0 ] 
  ..@ gxe    :List of 1
  .. ..$ : NULL
  ..@ fixEff : int [1:100] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ reps   : num [1:100] 1 1 1 1 1 1 1 1 1 1 ...
  ..@ nInd   : int 100
  ..@ nChr   : int 10
  ..@ ploidy : int 2
  ..@ nLoci  : int [1:10] 1000 1000 1000 1000 1000 1000 1000 1000 1000 1000
  ..@ geno   :List of 10
  .. ..$ : raw [1:125, 1:2, 1:100] ec 0e 6c 87 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 49 59 67 e7 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 43 60 4a 40 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 20 aa 43 e5 ...
  .. ..$ : raw [1:125, 1:2, 1:100] fa 80 b0 78 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 08 2c e0 45 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 52 f2 81 82 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 8e c8 88 14 ...
  .. ..$ : raw [1:125, 1:2, 1:100] c0 5b 13 a2 ...
  .. ..$ : raw [1:125, 1:2, 1:100] 16 e6 15 67 ...
  .. ..- attr(*, &quot;dim&quot;)= int [1:2] 10 1</code></pre>
</div>
<div id="population-information" class="section level2">
<h2>Population information</h2>
<p>The population has <code>id</code>s. The <code>@mother</code> and <code>@father</code> ids are all zero because this population was made from founder haplotypes, and so does not have diploid parents. The genotypic values <code>gv</code> of the population are calculated for the trait created using <code>SP$addTraitA(nQtlPerChr=nQTL)</code>. Given that there are 10 chromosomes and 10 QTL per chromosome, there are <code>3^(nChr*nQTL)</code> = 5.153775210^{47} different possible genotypic values. These genotypic values are accessible with the function <code>gv(founders)</code></p>
<p>From here, you can treat this population like a named vector using the square braces extraction operator <code>[ ]</code>. Extract individuals by their <code>@id</code> or just by their order in the population using an integer index. For example, pick three random individuals from a population and list their ids. Pick the one with the first id in alphabetical order.</p>
<pre class="r"><code>test &lt;- founders[sample(nFounders, 3)]
testID &lt;- test@id
firstInd &lt;- test[order(testID)[1]]
print(testID)</code></pre>
<pre><code>[1] &quot;5&quot;  &quot;94&quot; &quot;97&quot;</code></pre>
<pre class="r"><code>print(firstInd)</code></pre>
<pre><code>An object of class &quot;Pop&quot; 
Ploidy: 2 
Individuals: 1 
Chromosomes: 10 
Loci: 10000 
Traits: 1 </code></pre>
<pre class="r"><code>print(firstInd@id)</code></pre>
<pre><code>[1] &quot;5&quot;</code></pre>
</div>
<div id="outbreak-of-variation" class="section level2">
<h2>Outbreak of variation</h2>
<p>In the early 20th century, Emerson and East showed that if you crossed to inbreds, the hybrid had similar variation to each inbred, but if you then selfed the hybrid, the offspring varied substantially. This code simulates that result. First, self the founders to homozygosity. The function <code>self</code> self-fertilizes individuals from the population. By default, it creates one selfed individual per parent (controllable with the parameter <code>nProgeny</code>), so this works nicely for single-seed descent.</p>
<pre class="r"><code># Self-pollinate to for a few generations
nGenSelf &lt;- 4
inbredPop &lt;- founders
for (gen in 1:nGenSelf){
  inbredPop &lt;- self(inbredPop)
}</code></pre>
</div>
<div id="check-homozygosity" class="section level2">
<h2>Check homozygosity</h2>
<p>Just a sanity check that this has, in fact, created a population of 100 individuals that are appropriately homozygous. Loci are coded 0, 1, 2. So <code>qtl == 1</code> represents the case were a locus is heterozygous. <code>sum(qtl == 1)</code> counts those cases.</p>
<pre class="r"><code>qtl &lt;- pullQtlGeno(inbredPop)
if (nrow(qtl) != nFounders) stop(&quot;The number of individuals is unexpected&quot;)
if (ncol(qtl) != nChr * nQTL) stop(&quot;The number of QTL is unexpected&quot;)
fracHet &lt;- sum(qtl == 1) / (nFounders * nChr * nQTL)
cat(&quot;Expected fraction heterozygous&quot;, 1 / 2^nGenSelf, &quot;\n&quot;,
    &quot;Observed fraction heterozygous&quot;, fracHet, &quot;+/-&quot;, 
    round(2*sqrt(fracHet*(1-fracHet)/(nFounders*nChr*nQTL)), 3), &quot;\n&quot;)</code></pre>
<pre><code>Expected fraction heterozygous 0.0625 
 Observed fraction heterozygous 0.0162 +/- 0.003 </code></pre>
<blockquote>
<p>What was wrong with my reasoning about the Expected fraction heterozygous?</p>
</blockquote>
</div>
<div id="simulate-outbreak-of-variation" class="section level2">
<h2>Simulate outbreak of variation</h2>
<p>Just for fun, we will pick the pair of most average or most extreme inbred individuals, in terms of their genetic value, cross thpse pairs, then self-fertilize them to test their variation. We will assume a trait that has a heritability of 0.5 in the base, non-inbred population population. In practice, that means an error variance of 1. Here, I pick the average pair by finding the order and picking two in the middle. I pick the extreme pair by finding the order and picking the first and last.</p>
<pre class="r"><code>indGVs &lt;- gv(inbredPop)
averageInd &lt;- inbredPop[order(indGVs)[50:51]]
extremeInd &lt;- inbredPop[order(indGVs)[c(1, nFounders)]]</code></pre>
</div>
<div id="cross-the-pair-to-make-f1-hybrid" class="section level2">
<h2>Cross the pair to make F1 hybrid</h2>
<p>The crossPlan is a <strong>two-column matrix</strong> with as many rows as the number of crosses you want to make, the first column is the <code>@id</code> or the index of the seed parent, and likewise for the pollen parent in the second column. To make an F1, you just want to make one cross, so the matrix is one row. You just want to cross individual 1 with individual 2, so the row contains 1:2.</p>
<pre class="r"><code>crossPlan &lt;- matrix(1:2, nrow=1, ncol=2)
f1_average &lt;- makeCross(averageInd, crossPlan)
f1_extreme &lt;- makeCross(extremeInd, crossPlan)</code></pre>
</div>
<div id="make-f2-and-observe-phenotypic-variance" class="section level2">
<h2>Make F2 and observe phenotypic variance</h2>
<p>When you first make a population, <code>AlphaSimR</code> does not assume it has been phenotyped. You can phenotype it using the <code>setPheno</code> function. Note that if you use <code>setPheno</code> on the same population a second time, that will overwrite the phenotypes from the first time. I also print the genotypic variance that can be retrieved using the function <code>varG</code>. Really, <code>varG</code> gives all we need to know, but of course that variance is generally not observable in non-simulated reality.</p>
<pre class="r"><code>f2_average &lt;- self(f1_average, nProgeny=100)
f2_average &lt;- setPheno(f2_average, varE=1)
cat(&quot;Genotypic variance of F2 between average pair&quot;, round(varG(f2_average), 3))</code></pre>
<pre><code>Genotypic variance of F2 between average pair 0.346</code></pre>
<pre class="r"><code>hist(pheno(f2_average), main=&quot;Histogram of Average-Pair F2 Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab1_ProgrammingAlphaSimR.Rmd/Make%20F2-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-Make-F2-1">
Past versions of “Make F2-1.png”
</button>
</p>
<div id="fig-no-spaces-Make-F2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/d91a4ffd6c41d77658da6a9efc7d8f42c1e97555/docs/figure/Lab1_ProgrammingAlphaSimR.Rmd/Make F2-1.png" target="_blank">d91a4ff</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/602d7eb7d861d46218767fb5b678276a51545c4c/docs/figure/Lab1_ProgrammingAlphaSimR.Rmd/Make F2-1.png" target="_blank">602d7eb</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>f2_extreme &lt;- self(f1_extreme, nProgeny=100)
f2_extreme &lt;- setPheno(f2_extreme, varE=1)
cat(&quot;Genotypic variance of F2 between extreme pair&quot;, round(varG(f2_extreme), 3))</code></pre>
<pre><code>Genotypic variance of F2 between extreme pair 1.093</code></pre>
<pre class="r"><code>hist(pheno(f2_extreme), main=&quot;Histogram of Extreme-Pair F2 Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab1_ProgrammingAlphaSimR.Rmd/Make%20F2-2.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-Make-F2-2">
Past versions of “Make F2-2.png”
</button>
</p>
<div id="fig-no-spaces-Make-F2-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/d91a4ffd6c41d77658da6a9efc7d8f42c1e97555/docs/figure/Lab1_ProgrammingAlphaSimR.Rmd/Make F2-2.png" target="_blank">d91a4ff</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/602d7eb7d861d46218767fb5b678276a51545c4c/docs/figure/Lab1_ProgrammingAlphaSimR.Rmd/Make F2-2.png" target="_blank">602d7eb</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="observe-f1-phenotypic-variance" class="section level2">
<h2>Observe F1 phenotypic variance</h2>
<p>We only made a single F1 plant. <code>AlphaSimR</code> can’t give that plant different phenotypes. So to pretend to be Emerson and East, we have to make, say, 100 F1 individuals. I will do this just for the f1_extreme case. Here, varG may or may not be zero, depending on how many generations of selfing you did.</p>
<pre class="r"><code>crossPlan &lt;- matrix(1:2, nrow=100, ncol=2, byrow=T)
f1_extremeMany &lt;- makeCross(extremeInd, crossPlan)
f1_extremeMany &lt;- setPheno(f1_extremeMany, varE=1)
cat(&quot;Genotypic variance of F1 between extreme pair&quot;, round(varG(f1_extremeMany), 3))</code></pre>
<pre><code>Genotypic variance of F1 between extreme pair 0</code></pre>
<pre class="r"><code>hist(pheno(f1_extremeMany), main=&quot;Histogram of Extreme-Pair *F1* Phenotypes&quot;)</code></pre>
<p><img src="figure/Lab1_ProgrammingAlphaSimR.Rmd/Make%20many%20hybrids-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-no-spaces-Make-many-hybrids-1">
Past versions of “Make many hybrids-1.png”
</button>
</p>
<div id="fig-no-spaces-Make-many-hybrids-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/d91a4ffd6c41d77658da6a9efc7d8f42c1e97555/docs/figure/Lab1_ProgrammingAlphaSimR.Rmd/Make many hybrids-1.png" target="_blank">d91a4ff</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
</tr>
<tr>
<td>
<a href="https://github.com/jeanlucj/Lab1_7170/blob/602d7eb7d861d46218767fb5b678276a51545c4c/docs/figure/Lab1_ProgrammingAlphaSimR.Rmd/Make many hybrids-1.png" target="_blank">602d7eb</a>
</td>
<td>
jeanlucj
</td>
<td>
2020-01-20
</td>
</tr>
</tbody>
</table>
</div>
</div>
<blockquote>
<p>What conditions will tend to make the outbreak of variation under the “average” scenario similar to the “extreme” scenario?</p>
</blockquote>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.1 (2019-07-05)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Mojave 10.14.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.4.0    stringr_1.4.0    dplyr_0.8.3      purrr_0.3.3     
 [5] readr_1.3.1      tidyr_1.0.0      tibble_2.1.3     ggplot2_3.2.1   
 [9] tidyverse_1.3.0  AlphaSimR_0.11.0 R6_2.4.1         workflowr_1.6.0 

loaded via a namespace (and not attached):
 [1] tidyselect_0.2.5 xfun_0.11        haven_2.2.0      lattice_0.20-38 
 [5] colorspace_1.4-1 vctrs_0.2.1      generics_0.0.2   htmltools_0.4.0 
 [9] yaml_2.2.0       rlang_0.4.2      later_1.0.0      pillar_1.4.3    
[13] withr_2.1.2      glue_1.3.1       DBI_1.1.0        dbplyr_1.4.2    
[17] modelr_0.1.5     readxl_1.3.1     lifecycle_0.1.0  munsell_0.5.0   
[21] gtable_0.3.0     cellranger_1.1.0 rvest_0.3.5      evaluate_0.14   
[25] knitr_1.26       httpuv_1.5.2     fansi_0.4.0      broom_0.5.3     
[29] Rcpp_1.0.3       promises_1.1.0   backports_1.1.5  scales_1.1.0    
[33] jsonlite_1.6     fs_1.3.1         hms_0.5.2        digest_0.6.23   
[37] stringi_1.4.3    grid_3.6.1       rprojroot_1.3-2  cli_2.0.0       
[41] tools_3.6.1      magrittr_1.5     lazyeval_0.2.2   crayon_1.3.4    
[45] whisker_0.4      pkgconfig_2.0.3  zeallot_0.1.0    xml2_1.2.2      
[49] reprex_0.3.0     lubridate_1.7.4  rstudioapi_0.10  assertthat_0.2.1
[53] rmarkdown_2.0    httr_1.4.1       nlme_3.1-143     git2r_0.26.1    
[57] compiler_3.6.1  </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
